Refactor LLMRequest: Structured RequestData for Completions & Chat-Completions

vMaroon opened the pull request:
## **Motivation**

The types.LLMRequest struct is the core API for scheduling plugins.
Until now, it only exposed a flat Prompt string.
* For `/v1/completions`, this was sufficient.
* For `/v1/chat/completions`, messages were flattened into a pseudo-prompt via naive templating.

This flattening discards useful fields (e.g., `tools`, `chat_template`, etc.) that plugins may need. For example, the **llm-d** `precise-prefix-cache-scorer` plugin recreates vLLM's tokenization (including `jinja2` templating), but today's API does not provide all the necessary inputs.

To support richer scheduling logic and prepare for newer APIs (e.g., OpenAI responses), we need to preserve raw, structured request data instead of flattening it upfront.

## **Summary of Changes**

* **types.LLMRequest**
   * Replaced flat `Prompt` string with structured `LLMRequestData`, a disjoint union of `CompletionsRequest` and `ChatCompletionsRequest` - the latter includes fields from the HuggingFace transformers chat-templating API.
   * Request data is preserved as-is, **local** transformations are left to plugins.

* **Prefix-cache scorer**
   * No longer depends on naive prompt reconstruction for chat-completions:
     * This was originally done in order to attempt to hold the 1:4 chars-to-tokens estimation ratio. This attempt fails in that in actual chat-templating, the added special-keywords map to **individual preserved tokens** that are a part of the model's vocabulary. Therefore the existence of the naive templating as a whole is questionable, and can be avoided.

* **requestcontrol.Director**
   * Now responsible for parsing and populating the request fields as-is.
    
## **Testing**

* **Unit tests**
   * Added coverage for both completions and chat-completions request parsing.
   * Validated optional fields (tools, documents, chat_template, etc.).

* **Prefix plugin tests**
   * Verified prefix scoring works with both completions and chat-completions.
   * Added growth tests to confirm correct prefix matching across extended conversations.

* **Benchmarks**
   * Stress tests for long prompts and large chat histories.
   * Confirmed cache scoring scales correctly under load.

## **Related Issues**
* Fixes #827

cc @nirrozenbaum @kfswain @liu-cong

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin.go:
```diff
@@ -250,33 +250,37 @@ func (p *Plugin) matchLongestPrefix(ctx context.Context, hashes []BlockHash) map
+	res := make([]BlockHash, 0, 1+len(userInput)/cacheBlockSize)
 	// Add the model to the first block hash so that different models have different hashes even with the same body.
-
-	firstBlockSize := cacheBlockSize
```
Why changing line #267-273? This is a change of behavior. This block was actually added later to fix a bug that when the input is shorter than the cacheBlockSize, then all requests of the same model gets the same hash, effectively creating an "model affinity" effect.

vMaroon replied:
I'm not sure I caught that - to clarify:
1. From your description, that makes it an optimization and not a bug, right?
2. What does model-affinity mean for short requests? If a request is shorter than `cacheBlockSize` then scoring based on a hit is useless performance-wise, right? Then why should the **prefix-cache** scorer give any scoring on model affinity?
  3. What does model-affinity mean? An instance either has the model or does not. If it has it, then it can serve the request.

liu-cong replied:
> From your description, that makes it an optimization and not a bug, right?

This introduces a regression, see https://github.com/kubernetes-sigs/gateway-api-inference-extension/issues/1333 for context. 

The issue is, initially all short requests will be routed to the same pod until that pod gets overwhelmed (because the scorer will only consider the model name in the hash, and thus giving each request a score of 1)

vMaroon replied:
Fixed.

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin.go:
```diff
@@ -250,33 +250,39 @@ func (p *Plugin) matchLongestPrefix(ctx context.Context, hashes []BlockHash) map
-		block = append(block, toBytes(prevBlockHash)...)
-		res = append(res, BlockHash(xxhash.Sum64(block)))
+	h := xxhash.New()
+	_, _ = h.Write([]byte(request.TargetModel))
```
```suggestion
	h.Write([]byte(request.TargetModel))
```

vMaroon replied:
This breaks linting - reverted.

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin.go:
```diff
@@ -250,33 +250,39 @@ func (p *Plugin) matchLongestPrefix(ctx context.Context, hashes []BlockHash) map
+	prevBlockHash := BlockHash(h.Sum64())
+	for i := 0; i+cacheBlockSize <= len(userInput); i += cacheBlockSize {
+		h.Reset()
+		_, _ = h.Write(userInput[i : i+cacheBlockSize])
```
```suggestion
		h.Write(userInput[i : i+cacheBlockSize])
```

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin.go:
```diff
@@ -250,33 +250,39 @@ func (p *Plugin) matchLongestPrefix(ctx context.Context, hashes []BlockHash) map
+	for i := 0; i+cacheBlockSize <= len(userInput); i += cacheBlockSize {
+		h.Reset()
+		_, _ = h.Write(userInput[i : i+cacheBlockSize])
+		_, _ = h.Write(toBytes(prevBlockHash))
```
```suggestion
		h.Write(toBytes(prevBlockHash))
```

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin.go:
```diff
@@ -250,33 +250,39 @@ func (p *Plugin) matchLongestPrefix(ctx context.Context, hashes []BlockHash) map
 	// Split the body into blocks of size cacheBlockSize. The +1 is to account for the model.
 	// If the last block is smaller than cacheBlockSize, it will be ignored.
-	res := make([]BlockHash, 0, 1+len(prompt)/cacheBlockSize)
+	res := make([]BlockHash, 0, 1+len(userInput)/cacheBlockSize)
```
```suggestion
	res := make([]BlockHash, 0, len(userInput)/cacheBlockSize)
```

liu-cong replied:
Your implementation is slightly different than previous version, and the size of the `res` needs to be adjusted accordingly.

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin.go:
```diff
@@ -250,33 +250,39 @@ func (p *Plugin) matchLongestPrefix(ctx context.Context, hashes []BlockHash) map
+		loggerDebug.Info("Truncating input", "size", len(userInput), "max prefix blocks", maxPrefixBlocks, "block size", cacheBlockSize)
+		userInput = userInput[:maxPrefixBlocks*cacheBlockSize]
 	}
 	// Split the body into blocks of size cacheBlockSize. The +1 is to account for the model.
```
```suggestion
	// Split the body into blocks of size cacheBlockSize. 
```

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin_test.go:
```diff
@@ -34,7 +34,6 @@ import (
 )
 
 func TestPrefixPlugin(t *testing.T) {
```
```suggestion
func TestPrefixPluginCompletion(t *testing.T) {
```

liu-cong commented on pkg/epp/scheduling/framework/plugins/multi/prefix/plugin_test.go:
```diff
@@ -175,6 +194,149 @@ func TestPrefixPlugin(t *testing.T) {
 	plugin.PreRequest(context.Background(), req5, schedulingResult, 0)
 }
 
+func TestPrefixPluginChatCompletions(t *testing.T) {
```
nit: Looks like this test is unnecessary given the `TestPrefixPluginChatCompletionsGrowth` below

liu-cong commented on pkg/epp/scheduling/types/types.go:
```diff
@@ -29,14 +29,79 @@ type LLMRequest struct {
+	ChatCompletions *ChatCompletionsRequest `json:"chat_completions,omitempty"`
+}
+
+func (r *LLMRequestData) String() string {
```
nit: I don't think this is needed given you have String method overrides for individual fields.

liu-cong commented on pkg/epp/scheduling/types/types.go:
```diff
@@ -29,14 +29,79 @@ type LLMRequest struct {
 func (r *LLMRequest) String() string {
-	return fmt.Sprintf("RequestID: %s, TargetModel: %s, PromptLength: %d, Headers: %v", r.RequestId, r.TargetModel, len(r.Prompt), r.Headers)
+	return fmt.Sprintf("RequestID: %s, TargetModel: %s, RequestData: %s, Headers: %v",
+		r.RequestId, r.TargetModel, r.Data.String(), r.Headers)
```
```suggestion
		r.RequestId, r.TargetModel, r.Data, r.Headers)
```

liu-cong commented on pkg/epp/scheduling/types/types.go:
```diff
@@ -29,14 +29,79 @@ type LLMRequest struct {
+	Prompt string
+}
+
+func (r *CompletionsRequest) String() string {
```
nit: I suggest adding `if r==nil {return ""}` which is safer and also removed the need to add a String function on the parent struct.

vMaroon replied:
Returning "<nil>" instead of empty for conciseness. On removing the parent struct's override, I rather keep it for better log readability.

liu-cong commented on pkg/epp/scheduling/types/types.go:
```diff
@@ -29,14 +29,79 @@ type LLMRequest struct {
+	ChatTemplateKWArgs        map[string]interface{} `json:"chat_template_kwargs,omitempty"`
+}
+
+func (r *ChatCompletionsRequest) String() string {
```
same here, check `r == nil`

liu-cong commented on pkg/epp/scheduling/types/types.go:
```diff
@@ -29,14 +29,79 @@ type LLMRequest struct {
+// API spec.
+type ChatCompletionsRequest struct {
+	/* parameters from the official OpenAI chat-completions API */
+	Messages []Message
```
do we need to add the `json:"messages"` ?

liu-cong commented on pkg/epp/util/request/body.go:
```diff
@@ -17,70 +17,52 @@ limitations under the License.
+	return &types.LLMRequestData{ChatCompletions: &chatCompletions}, nil
+}
+
+func validateChatCompletionsMessages(messages []types.Message) error {
```
If there a reason we must do the validation here? If not I'd rather defer this to the model server

liu-cong commented on pkg/epp/util/request/body_test.go:
```diff
@@ -39,169 +52,369 @@ func TestExtractPromptFromRequestBody(t *testing.T) {
+	return true
+}
+
+func compareSliceAny(got, want []any) bool {
```
any reason we need a custom compare instead of `cmp.Diff`? Same comment for the map comparison

kfswain commented:
/hold

While we are releasing

vMaroon commented:
@liu-cong thank you for reviewing - addressed your comments. I force pushed a squash with 3 of your commits due to the problem above - hope that is ok.

liu-cong reviewed (commented):
/lgtm 

just a nit

vMaroon commented:
Bumping since the release was cut @kfswain @nirrozenbaum @liu-cong

ahg-g commented on pkg/epp/requestcontrol/director.go:
```diff
@@ -121,7 +122,7 @@ func (d *Director) HandleRequest(ctx context.Context, reqCtx *handlers.RequestCo
 		RequestId:   reqCtx.Request.Headers[requtil.RequestIdHeaderKey],
 		TargetModel: reqCtx.TargetModelName,
-		Prompt:      prompt,
+		Data:        requestData,
```
nit: data is a generic name, is there a more specific name we can use here? should we use body instead since we have headers below?

vMaroon replied:
`Input`?

ahg-g replied:
isn't this the body but structured?

ahg-g replied:
In LLMRequest, it is the only part of the body being captured, so I think Body is appropriate here; I think the type `LLMRequestData` could be named `Body` as well; if we want to pass the raw body in the future, we can add a `RawBody` field later

ahg-g commented on pkg/epp/util/request/body.go:
```diff
@@ -17,70 +17,43 @@ limitations under the License.
+// ExtractRequestData extracts the LLMRequestData from the given request body map.
+func ExtractRequestData(body map[string]any) (*types.LLMRequestData, error) {
+	// Convert map back to JSON bytes
+	jsonBytes, err := json.Marshal(body)
```
Is there a perf concern here that we are marshalling and unmarshalling again for every request? What are your thoughts on having the [Body param in the Request type](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/epp/handlers/server.go#L111) being structured to begin with so that we unmarshal once?

vMaroon replied:
This PR initially began with extracting each field separately by checking the body map, but this change was suggested in this comment https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/1446#discussion_r2299148093 for better readability.

I think your proposal makes sense but it could get complex due to the different APIs and additional fields that the different inference-engines might add. I think this should be a follow-up discussion/work, outside of this PR.

FYI, in a benchmark that compares chat-completions with the entire e2e pipeline of this change, plus the llm-d precise-prefix-cache with preprocessing (using an embedded python interpreter), added about 10ms E2E latency compared to llm-d-inference-scheduler v0.2.1 on completions API.
